kernel: sched: default TIMESLICE_SIZE to 20

hasheddan opened the pull request:
Sets the default for CONFIG_TIMESLICE_SIZE to 20 to match the historical clamping behavior that was removed in [405611d](https://github.com/zephyrproject-rtos/zephyr/commit/405611dc9e0edb93b6f05bbc1597024fe34be0b7).

Fixes #89965

hasheddan committed:
kernel: sched: default TIMESLICE_SIZE to 20

Sets the default for CONFIG_TIMESLICE_SIZE to 20 to approximate the
historical clamping behavior that was removed in 405611d.

Signed-off-by: Daniel Mangum <georgedanielmangum@gmail.com>

andyross reviewed (dismissed):
Love it, thanks

jameswalmsley commented on kernel/Kconfig:
```diff
@@ -653,7 +653,7 @@ config TIMESLICING
-	default 0
+	default 2
 	range 0 $(INT32_MAX)
 	depends on TIMESLICING
```
Appreciate the default was set to 2 because of the historical clamping, but wouldn't the sensible default be 1ms?

I think most people would expect time-slices of 1ms, especially when having used other RTOS's like FreeRTOS in the past.

andyross replied:
Just history and compatibility.  Zero has always been legal (as a synonym for CONFIG_TIMESLICING=n, I guess).  No reason not to have discussions about API changes, but this bug was a default behavior regression.

cfriedt commented:
I'm a bit worried about the failing tests. Are those related to the linked issue? If so, we should probably add some commits to fix them.

nashif commented:
> I'm a bit worried about the failing tests. Are those related to the linked issue?

they are related. slicing is enabled by default, and now with the TIMESLICE_SIZE  being set to 2, so tests need to be adapted if they were not expecting slicing.

hasheddan commented:
> they are related. slicing is enabled by default, and now with the TIMESLICE_SIZE being set to 2, so tests need to be adapted if they were not expecting slicing.

@nashif I'm happy to update those in this PR or do so in a follow up

hasheddan commented:
@nashif great -- for each of the failing tests the simplest update is just `CONFIG_TIMESLICING=n`.I do think it would be good to add variations for some of these tests with timeslicing enabled. However, given that these tests were written without timeslicing in mind, I think the smallest diff (i.e. `CONFIG_TIMESLICING=n`) may be the best option for this PR. What are your thoughts?

npitre reviewed (commented):
In fact the historical clamping was for 2 _ticks_ not 2 millisecs.
So having a default of 2 millisecs here does not exactly relate to
anything (except for those configs for which it is equivalent).

Therefore I'd lean towards having a default of 1 instead, as already
suggested. The actual fix is for that default to not be zero, not about
being 2.

andyross commented:
LOL, I completely forgot that timeslicing was an older API and still exposed ms to the user.  So yeah, no magic to "2", just anything non-zero.  When originally written, ticks rate was 100 Hz by default, so "20" might be closer to the original design intent I guess.  But now tick rate varies widely between platforms and "2 ticks" might be anything from 20ms (qemu) down to ~90us (nRF).

As far as how to adjust the tests, best way to avoid changing behavior would be to leave TIMESLICING=y and set TIMESLICE_SIZE=0.  The former disables the feature entirely and changes the generated code, even if the visible behavior should be the same.

andyross reviewed (dismissed):
Refresh +1 regardless, those tests don't exercise timeslicing anyway

hasheddan commented:
It appears `20` ticks has been used as a pseudo default in some other places (e.g. https://github.com/zephyrproject-rtos/zephyr/commit/f08c9fe2c2b3a34d87236ad3d47be75db86cef46).

> As far as how to adjust the tests, best way to avoid changing behavior would be to leave TIMESLICING=y and set TIMESLICE_SIZE=0. The former disables the feature entirely and changes the generated code, even if the visible behavior should be the same.

Good point -- will switch to that strategy 👍🏻

hasheddan committed:
tests: kernel: set timeslice size to 0 in tests that do not exercise it

Sets CONFIG_TIMESLICE_SIZE=0 for tests that previously expected to run
without timeslicing enabled, and now fail with the default
CONFIG_TIMESLICE_SIZE=20. Note that timeslicing functionality is still
included, but no actual timeslicing is performed.

Signed-off-by: Daniel Mangum <georgedanielmangum@gmail.com>

andyross reviewed (approved):
Still +1

kartben commented:
Main is extremely unhappy about the change. Should we revert it? Or do more tests need updating?

hasheddan commented:
@kartben I opened up a follow up in #90093 to address 👍🏻
